packages/experimental/Avatar/src/Avatar.tsx

```diff
@@ -0,0 +1,176 @@
+   * Supported Avatar sizes
+   */
+  size?: Size;
+
+  /**
+   * Background Color of initials
+   */
+  backgroundColor?: string;
+
+  /**
+   * Image to be used as border around the avatar. It will be used as a pattern image color,
+   * but It will be scaled to fit the avatar size. If set, the hasBorder initializer value will be ignored,
+   * since it's assumed that the client intends to have a custom border.
+   */
+  customBorderImage?: Image;
+
+  /**
+   * Shape of the AvatarView
+   * Circle is used to represent people
+   * Square is used to represent organizations or teams
+   */
+  avatarStyle?: AvatarStyle;
+
+  /**
+   * The color of the border of the avatar view
+   */
+  borderColor?: string;
+
+  /**
+   * When true, the presence status border is opaque. Otherwise, it is transparent.
+   */
+  usesOpaquePresenceBorder?: boolean;
+
+  // TODO Github#512, Map the viewProp AccessibilityLabel to this
+  // overrideAccessibilityLabel? : string;
+
+  /**
+   * Used when avatarView doesn't have image or can't generate initials string
+   */
+  preferredFallBackImageStyle?: string;
+
+  /**
+   * Set to true to enable the pointer interaction on the avatar view, false by default.
+   * TODO Figure out how to make iOS only
+   */
+  hasPointerInteraction?: boolean;
+};
+
+export type AvatarTokens = {
+  /**
+   * Supported Avatar sizes
+   */
+  size?: Size;
+};
+
+const tokensThatAreAlsoProps: (keyof AvatarTokens)[] = ['size'];
+
+// The Javascript API is a flat list for all the props we can set on our component
+export type AvatarProps = ViewProps & otherAvatarProps & AvatarData;
+
+// The Native component API has a sub object that we flattened out in AvatarProps
+export type NativeAvatarViewProps = ViewProps &
+  otherAvatarProps & {
+    avatarData?: AvatarData;
+  };
+
+interface AvatarType {
+  props: AvatarProps;
+  slotProps: { root: NativeAvatarViewProps };
+  tokens: AvatarTokens;
+}
+
+export const Avatar = compose<AvatarType>({
+  displayName: avatarName,
+  tokens: [
+    {
+      size: 'small',
+    },
+    avatarName,
+  ],
+  tokensThatAreAlsoProps,
+  slots: { root: NativeAvatarView },
+  slotProps: {
+    root: buildProps<NativeAvatarViewProps, AvatarTokens>(
+      (tokens) => ({
+        size: tokens.size,
+        style: {
+          height: ExportedNativeConstants.sizes[tokens.size],
+          width: ExportedNativeConstants.sizes[tokens.size],
+        },
+      }),
+      ['size'],
+    ),
+  },
+  render: (props: AvatarProps, useSlots: UseSlots<AvatarType>) => {
+    const Root = useSlots(props).root;
+
+    const cachedAvatarDataProp = cache(
+      () => ({
+        primaryText: props.primaryText,
+        secondaryText: props.secondaryText,
+        image: props.image,
+        color: props.color,
+        customBorderImage: props.customBorderImage,
+        presence: props.presence,
+      }),
+      [props.primaryText, props.secondaryText, props.image, props.color, props.customBorderImage, props.presence],
```

Comment:
One thing to consider when caching is the likelihood of re-use vs memory overhead. Given that this data includes content values, and as such every element of a list might have a different avatar data, your primary goal will be to insure render to render consistency, rather than cross-component sharing. From this I would switch this from memo-cache to React.useMemo.  Here's a quick breakdown of when to use each:

## memo-cache
**How it works:** Persistent, hierarchical (per-key) cache, scope of storage is up to you as the user controls the cache reference.
**Best uses**: When data can often be shared across component instances. E.g. if you are rendering a hundred instances on a screen and the expectation is that much of the data can be shared across component instances.

## React.useMemo
**How it works**: Actually the baseline for almost all hook implementations (useRef, useCallback, etc.) This stores the value in a hook slot (just like useState would), and compares the keys via a mixed array. This will not share across components, and only stores the last used. But it is lighter weight and likely a touch faster.
**When to use**: When you just want to ensure that if you render a single component twice in a row, the sub-objects don't mutate. When the values to memoize involve content (like a label) that will be different for every component.

Reply:
Thanks for the explanation! I've switched to useMemo